driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -337,30 +289,49 @@ private PoolingOptions options() {
       }
     }
 
-    if (connections[shardId].isEmpty()) {
-      if (host.convictionPolicy.canReconnectNow()) {
-        if (connectionsPerShard == 0) {
-          maybeSpawnNewConnection(shardId);
-        } else if (scheduledForCreation[shardId].compareAndSet(0, connectionsPerShard)) {
-          for (int i = 0; i < connectionsPerShard; i++) {
-            // We don't respect MAX_SIMULTANEOUS_CREATION here because it's  only to
-            // protect against creating connection in excess of core too quickly
-            manager.blockingExecutor().submit(new ConnectionTask(shardId));
-          }
-        }
-        return enqueue(timeout, unit, maxQueueSize, shardId);
-      }
+    int index = 0;
```

Comment:
This selection of connection to different shard does not seem right. This way we may never try to reconnect to a shard that loses all its connections.

Reply:
The previous version of the code was making sure that for each shard if there wasn't any connection it would try to create a new one. After this change, the code will just ignore shards that have no connections and instead take connections to different shards.

Reply:
No, the pool will keep opening connections until all shards are full. Review the addCallback method. You're looking at the wrong spot.

Reply:
addCallback is called only on initialization. This method can be called much later than that. Imagine a case when the initialization has successfully finished and then for some reason one of the connections gets dropped. In such a case, borrowConnection should try to recreate such connection. In general, it will but if all connections for a single shard get dropped it won't consider it anymore and will keep using connections to other shards instead.